### PR TITLE
Add support and document how to expose ssh port in dockerized gitlab-ce

### PR DESCRIPTION
**Pull request by harobed**
Hi,

the goal of this Pull Request is to add support and document how to expose ssh port in dockerized gitlab-ce.

To implement that, I needed to add `%h/.ssh/authorized_keys_proxy` to `AuthorizedKeysFile` parameter in `/etc/ssh/sshd_config` to support ssh host forwarding to GitLab container.

How to test it: https://github.com/harobed/docker-gitlab-vagrant-test

### CODE REVIEW

File: contrib/expose-gitlab-ssh-port.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -ev
+
+if ! id -u git >> /dev/null 2>&1; then
+  groupadd -g 1010 git
```
Comment (solidnerd) [inline contrib/expose-gitlab-ssh-port.sh]: I think we should make the group id changeable via variable or a flag 
```
GITLAB_USERGROUP=$1
GITLAB_USERGROUP=${GITLAB_USERGROUP:-1010}
```

File: contrib/expose-gitlab-ssh-port.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  useradd -m -u 1010 -g git -s /bin/sh -d /home/git git
```
Comment (solidnerd) [inline contrib/expose-gitlab-ssh-port.sh]: The same here for the user

File: contrib/expose-gitlab-ssh-port.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+su git -c "if [ ! -f /home/git/.ssh/id_rsa ]; then ssh-keygen -t rsa -N \"\" -f /home/git/.ssh/id_rsa; fi"
+su git -c "if [ -f /home/git/.ssh/id_rsa.pub ]; then mv /home/git/.ssh/id_rsa.pub /home/git/.ssh/authorized_keys_proxy; fi"
+
+mkdir -p /home/git/gitlab-shell/bin/
+rm -f /home/git/gitlab-shell/bin/gitlab-shell
+tee -a /home/git/gitlab-shell/bin/gitlab-shell > /dev/null <<'EOF'
+#!/bin/sh
+
+ssh -i /home/git/.ssh/id_rsa -p 9922 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $0 $@"
```
Comment (solidnerd) [inline contrib/expose-gitlab-ssh-port.sh]: Here also the ssh port

File: contrib/expose-gitlab-ssh-port.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -ev
+
+if ! id -u git >> /dev/null 2>&1; then
+  groupadd -g 1010 git
+  useradd -m -u 1010 -g git -s /bin/sh -d /home/git git
+fi
+su git -c "mkdir -p /home/git/.ssh/"
+
+su git -c "if [ ! -f /home/git/.ssh/id_rsa ]; then ssh-keygen -t rsa -N \"\" -f /home/git/.ssh/id_rsa; fi"
```
Comment (solidnerd) [inline contrib/expose-gitlab-ssh-port.sh]: generate the rsa key with a least 4096 bits to be a little more secure

File: README.md [hunk 5]
```diff
@@ -801,6 +802,16 @@ These options should contain something like:
 - `PIWIK_URL=piwik.example.org`
 - `PIWIK_SITE_ID=42`
 
+### Exposing ssh port in dockerized gitlab-ce
```
Comment (solidnerd) [inline README.md]: I think we should have a separate document for this and linking only here to document.

So than we could have a documented in the docs. 

As an example have a look on https://github.com/sameersbn/docker-gitlab/blob/master/docs/container_registry.md or https://github.com/sameersbn/docker-gitlab/blob/master/docs/s3_compatible_storage.md

File: contrib/expose-gitlab-ssh-port.sh (hunk 1 above)
Comment (harobed) [reply contrib/expose-gitlab-ssh-port.sh]: 👍

**Comment by harobed** [comment]
@solidnerd thanks for your code review, I have resolved all your conversation.

Can you review my last update?

File: assets/build/install.sh [hunk 6]
```diff
@@ -210,6 +210,7 @@ sed -i \
   -e "s|^[#]*UsePrivilegeSeparation yes|UsePrivilegeSeparation no|" \
   -e "s|^[#]*PasswordAuthentication yes|PasswordAuthentication no|" \
   -e "s|^[#]*LogLevel INFO|LogLevel VERBOSE|" \
+  -e "s|^[#]*AuthorizedKeysFile.*|AuthorizedKeysFile %h/.ssh/authorized_keys %h/.ssh/authorized_keys_proxy|" \
```
[the remark this answers is not in the document]
Comment (harobed) [reply assets/build/install.sh]: > Probably we don't to add this to the install.sh

I have not noticed `%h/.ssh/authorized_keys_proxy` undesirable side effects.

[omnibus GitLab Docker Image](https://gitlab.com/gitlab-org/omnibus-gitlab/commit/923fd761ed854ca368c413a581b1153cd677dbe5#d99621b50fd86a7df71d4fa8a395a3736e43fc90) have also this hack.

> And mount the config back inside the container because it was created during build and changed by the runtime

It was my first implementation but I can't use it, I had a error (I do not remember precisely the error name) 😞.

Do you want that I spend some time to retrieve the error?

**Comment by harobed** [comment]
> But in terms of security it's a big hack.

Why? Where?

**Comment by harobed** [comment]
@solidnerd My last comments are clear?

File: contrib/expose-gitlab-ssh-port.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  useradd -m -u ${GITLAB_USERGROUP} -g git -s /bin/sh -d /home/git git
+fi
+su git -c "mkdir -p /home/git/.ssh/"
+
+su git -c "if [ ! -f /home/git/.ssh/id_rsa ]; then ssh-keygen -t rsa -b 4096 -N \"\" -f /home/git/.ssh/id_rsa; fi"
+su git -c "if [ -f /home/git/.ssh/id_rsa.pub ]; then mv /home/git/.ssh/id_rsa.pub /home/git/.ssh/authorized_keys_proxy; fi"
+
+mkdir -p /home/git/gitlab-shell/bin/
+rm -f /home/git/gitlab-shell/bin/gitlab-shell
+tee -a /home/git/gitlab-shell/bin/gitlab-shell > /dev/null <<EOF
```
Comment (ccoenen) [inline contrib/expose-gitlab-ssh-port.sh]: `tee ... > /dev/null` looks like a leftover from debugging, is that right?

Comment (harobed) [reply contrib/expose-gitlab-ssh-port.sh]: I do not remember what I use `tee`, maybe because that https://github.com/koalaman/shellcheck/wiki/SC2024

> tee ... > /dev/null looks like a leftover from debugging, is that right?

@ccoenen do you want to replace that by:

```
cat <<EOF > /home/git/gitlab-shell/bin/gitlab-shell
...
...
EOF
```

isn't it?

**Comment by harobed** [comment]
@solidnerd, the gentlest of bumps on this. Have two minute to take a look on this review?

**Comment by ccoenen** [comment]
I currently have a setup that runs on https only, because I need SSH for the host system (and I have no control over the firewall, so I can't just remap to a different port either).

This sounds like it is the perfect solution. I am hesitant to run this unless someone more familiar with gitlab or SSH has taken a closer look at it. Could someone give this a blessing or at least comment on it?

**Comment by harobed** [comment]
@solidnerd I have seen and merged all your [Pull Requests](https://github.com/harobed/docker-gitlab-vagrant-test/pulls?q=is%3Apr+is%3Aclosed) in [docker-gitlab-vagrant-test](https://github.com/harobed/docker-gitlab-vagrant-test) repository.

I assume that you have tested `expose-gitlab-ssh-port.sh` feature? If yes, it works for you?

**Comment by harobed** [comment]
>  Could someone give this a blessing or at least comment on it?

@ccoenen 

* I have followed this post https://blog.xiaket.org/2017/exposing.ssh.port.in.dockerized.gitlab-ce.html to write [expose-gitlab-ssh-port.sh](https://github.com/sameersbn/docker-gitlab/blob/5a477cbb373bfcb7a22e84a4ad4ed81c8ac420a9/contrib/expose-gitlab-ssh-port.sh)
* you can test it in this test environment [docker-gitlab-vagrant-test](https://github.com/harobed/docker-gitlab-vagrant-test)

I'd be happy to get your feedback 🙂

**Comment by solidnerd** [comment]
Hey @harobed,

yes I tested it it works. It found some configuration edge cases but I think this is out of scope. The method works only if you don't enable fast lookup for authorized ssh keys.

https://docs.gitlab.com/ce/administration/operations/fast_ssh_key_lookup.html .


For the first time we could add this. 

So LGTM !

**Comment by harobed** [comment]
@solidnerd thanks 🙂
